Bottom: 14c6baaec41790b80618cd1fd31f6bc28e66e638
Top:    9c1a068dfc0722f21ba326c851e0cd8af101e383
Author: Peter Clifton <pcjc2@cam.ac.uk>
Date:   2012-01-12 22:30:35 +0000

hid/gtk: Add feature to force an immediate redraw if framerate is low

If more than a certain time (currently hard-coded at 50ms) has elapsed
since the last expose callback, invalidate_all() will wait and process
updates before returning.

This is handy to allow animation of auto-router progress, where the gui
mainloop is not hit until the auto-routing operation is fully complete.


---

diff --git a/src/hid/gtk/gtkhid-gl.c b/src/hid/gtk/gtkhid-gl.c
index f3133ec..5da6a1a 100644
--- a/src/hid/gtk/gtkhid-gl.c
+++ b/src/hid/gtk/gtkhid-gl.c
@@ -48,6 +48,7 @@ typedef struct render_priv {
   int subcomposite_stencil_bit;
   char *current_colorname;
   double current_alpha_mult;
+  GTimer *time_since_expose;
 
   /* Feature for leading the user to a particular location */
   guint lead_user_timeout;
@@ -609,10 +610,17 @@ ghid_invalidate_lr (int left, int right, int top, int bottom)
   ghid_invalidate_all ();
 }
 
+#define MAX_ELAPSED (50. / 1000.) /* 50ms */
 void
 ghid_invalidate_all ()
 {
+  render_priv *priv = gport->render_priv;
+  double elapsed = g_timer_elapsed (priv->time_since_expose, NULL);
+
   ghid_draw_area_update (gport, NULL);
+
+  if (elapsed > MAX_ELAPSED)
+    gdk_window_process_all_updates ();
 }
 
 void
@@ -770,6 +778,8 @@ ghid_init_renderer (int *argc, char ***argv, GHidPort *port)
 
   port->render_priv = priv = g_new0 (render_priv, 1);
 
+  priv->time_since_expose = g_timer_new ();
+
   gtk_gl_init(argc, argv);
 
   /* setup GL-context */
@@ -956,6 +966,8 @@ ghid_drawing_area_expose_cb (GtkWidget *widget,
 
   ghid_end_drawing (port);
 
+  g_timer_start (priv->time_since_expose);
+
   return FALSE;
 }
